docs/designs/topology.md

```diff
@@ -0,0 +1,916 @@
+# Topology-Aware Scheduling - Grove Operator Design
+
+## Overview
+
+This document defines the design for implementing topology-aware scheduling in the Grove operator.
+
+**Motivation**: Topology-aware scheduling is critical for Grove's multinode inference workloads because these
+applications require:
+
+- **Network Locality**: Proximity improves high-bandwidth communication between leaders and their respective workers (
+  prefill and decode, etc)
+    - **Coordinated Placement**: Related components (e.g., model shards) perform better when co-located within the same
+      topology domain
+  - **Latency Optimization**: Minimizing network hops between interdependent inference components improves end-to-ends
+      performance
+
+**Design Approach**: This design introduces a flexible topology system with three main components:
+
+1. **TopologyDomain CRD**: Admin-configured cluster topology hierarchy mapping friendly names to node labels
+    2. **Operator Configuration**: Selects active topology via `--topology-domain-name` argument
+    3. **TopologyConstraint**: User-specified packing requirements in workloads (PodCliqueSet, PodCliqueScalingGroup,
+       PodClique)
+
+**Key Feature**: Grove provides automatic out-of-box topology optimization by generating preferred packing constraints
+at all levels, even without user configuration. Users can optionally specify required constraints for strict placement
+requirements.
+
+## Goals
+
+- Provide flexible, cluster-agnostic topology hierarchy definition via TopologyDomain CRD
+    - Enable packing constraints for network locality across all Grove scalable resources
+    - Support multiple topology configurations for different environments
+    - Automatic Kueue Topology generation for KAI scheduler integration
+    - Immutable topology configuration ensuring scheduling consistency
```

Comment:
Any reason to make it `immutable`?
Since we support scale-out it is quite possible that the scale-out now does not fit the resource pack constraints w.r.t available capacity. In this case would you not allow the user to relax the pack constraints?